Restructure of lightning logging docs (#907)
PyTorchLightning/metrics: docs/source/pages/lightning.rst

TorchMetrics in PyTorch Lightning
#################################

TorchMetrics was originally created as part of `PyTorch Lightning <https://github.com/PyTorchLightning/pytorch-lightning>`_, a powerful deep learning research
framework designed for scaling models without boilerplate.

.. note::

    TorchMetrics always offers compatibility with the last 2 major PyTorch Lightning versions, but we recommend to always keep both frameworks
    up-to-date for the best experience.

While TorchMetrics was built to be used with native PyTorch, using TorchMetrics with Lightning offers additional benefits:

> * Module metrics are automatically placed on the correct device when properly defined inside a LightningModule.

Reviewer: ```suggestion
* Modular metrics are automatically placed on the correct device when properly defined inside a LightningModule.
```

>   This means that your data will always be placed on the same device as your metrics. No need to call `.to(device)` anymore!

Reviewer: ```suggestion
  This means that your data will always be placed on the same device as your metrics. No need to call ``.to(device)`` anymore!
```

            # log epoch metric
            self.log('train_acc_epoch', self.accuracy)

> Metric logging in Lightning happens through the `self.log` or `self.log_dict` method. Both methods only supports logging of *scalar-tensors*.

Reviewer: ```suggestion
Metric logging in Lightning happens through the ``self.log`` or ``self.log_dict`` method. Both methods only support the logging of *scalar-tensors*.
```

> While the vast majority of metrics in torchmetrics returns a scalar tensor, some metrics such as :class:`~torchmetrics.ConfusionMatrix`, :class:`~torchmetrics.ROC`,

Reviewer 2: These refs looks outdated

Author: should be updated now :]

> While the vast majority of metrics in torchmetrics returns a scalar tensor, some metrics such as :class:`~torchmetrics.ConfusionMatrix`, :class:`~torchmetrics.ROC`,
> :class:`~torchmetrics.MAP`, :class:`~torchmetrics.RougeScore` return outputs that are non-scalar tensors (often dicts or list of tensors) and should therefore be

Author: ```suggestion
While the vast majority of metrics in torchmetrics returns a scalar tensor, some metrics such as :class:`~torchmetrics.ConfusionMatrix`, :class:`~torchmetrics.ROC`,
:class:`~torchmetrics.MeanAveragePrecision`, :class:`~torchmetrics.ROUGEScore` return outputs that are non-scalar tensors (often dicts or list of tensors) and should therefore be
```

dealt with separately. For info about the return type and shape please look at the documentation for the ``compute`` method for each metric you want to log.

********************
Logging TorchMetrics
********************

> Logging metrics can be done in two ways: either logging the metric object directly or the computed metric values. When :class:`~torchmetrics.Metric` objects, which return a scalar tensor,
> are logged directly in Lightning using the LightningModule `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ method.

Reviewer: ```suggestion
Logging metrics can be done in two ways: either logging the metric object directly or the computed metric values. When :class:`~torchmetrics.Metric` objects, which return a scalar tensor
are logged directly in Lightning using the LightningModule `self.log <https://pytorch-lightning.readthedocs.io/en/stable/extensions/logging.html#logging-from-a-lightningmodule>`_ method,
```

            self.valid_acc(logits, y)
            self.log('valid_acc', self.valid_acc, on_step=True, on_epoch=True)

As an alternative to logging the metric object and letting Lightning take care of when to reset the metric etc. you can also manually log the output
of the metrics.

.. testcode:: python

    class MyModule(LightningModule):

        def __init__(self):
            ...
            self.train_acc = torchmetrics.Accuracy()
            self.valid_acc = torchmetrics.Accuracy()

        def training_step(self, batch, batch_idx):
            x, y = batch
            preds = self(x)
            ...
            batch_value = self.train_acc(preds, y)
            self.log('train_acc_step', batch_value)

        def training_epoch_end(self, outputs):
            self.train_acc.reset()

        def validation_step(self, batch, batch_idx):
            logits = self(x)
            ...
            self.valid_acc.update(logits, y)

        def validation_epoch_end(self, outputs):
            self.log('valid_acc_epoch', self.valid_acc.compute())
            self.valid_acc.reset()

> Note that logging metrics this ways will require you to manually reset the metrics at the end of the epoch yourself. In general we recommend logging
> the metric object to make sure that metrics are correctly computed and reset. Additionally, we highly recommend that the two ways of logging is not
> mixed as it can lead to wrong results.

Reviewer: ```suggestion
Note that logging metrics this way will require you to manually reset the metrics at the end of the epoch yourself. In general, we recommend logging
the metric object to make sure that metrics are correctly computed and reset. Additionally, we highly recommend that the two ways of logging are not
mixed as it can lead to wrong results.
```

.. note::

    When using any Modular metric, calling ``self.metric(...)`` or ``self.metric.forward(...)`` serves the dual purpose of calling ``self.metric.update()``
    on its input and simultaneously returning the metric value over the provided input. So if you are logging a metric *only* on epoch-level (as in the
    example above), it is recommended to call ``self.metric.update()`` directly to avoid the extra computation.

    .. testcode:: python

        class MyModule(LightningModule):

            def __init__(self):
                ...
                self.valid_acc = torchmetrics.Accuracy()

            def validation_step(self, batch, batch_idx):
                logits = self(x)
                ...
                self.valid_acc.update(logits, y)
                self.log('valid_acc', self.valid_acc, on_step=True, on_epoch=True)

***************

> Common pitfalls

Reviewer: ```suggestion
Common Pitfalls
```

***************

The following contains a list of pitfalls to be aware of:

* If using metrics in data parallel mode (dp), the metric update/logging should be done
  in the ``<mode>_step_end`` method (where ``<mode>`` is either ``training``, ``validation``

>   or ``test``). This is due to metric states else being destroyed after each forward pass,
>   leading to wrong accumulation. In practice do the following:

Reviewer: ```suggestion
  or ``test``). This is because ``dp`` split the batches during the forward pass and metric states are destroyed after each forward pass, thus leading to wrong accumulation. In practice do the following:
```

.. testcode:: python

    class MyModule(LightningModule):

        def training_step(self, batch, batch_idx):
            data, target = batch
            preds = self(data)
            # ...
            return {'loss': loss, 'preds': preds, 'target': target}

        def training_step_end(self, outputs):

>             #update and log

Reviewer: ```suggestion
            # update and log
```
